# Boucles en Java: Algorithmique

Dans le notebook précédent on a vu comment parcourir un intervalle de nombres, et afficher les nombres.

On a aussi vu des boucles conditionnelles très simples.

Ici on va voir des cas d'utilisation plus pratiques des boucles inconditionnelles: trouver un élément, compter des éléments, faire une somme ou un produit.

## Trouver un élément dans une liste

Le problème général de trouver un élément dans une liste est très répandu, et sujet à une erreur très classique.

Pour l'instant, comme "liste" on va simplement prendre un intervalle de nombres, qu'on sait maintenant parcourir avec une boucle ```for```.

**Question** : Y a-t-il un multiple de 41 dans l'intervalle \[191, 207\]?

Pour répondre à cette question, on va parcourir l'intervalle et chercher un nombre qui répond à cette condition.

La boucle for pour parcourir l'intervalle (on va juste afficher les nombres):

In [6]:
for (int i=191; i<=207; i++){
    System.out.println(i);
}

191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207


Ok, on a maintenant la liste de nombres. Maintenant dans cette liste on va voir s'il y a un multiple de 43. Pour savoir si un nombre est multiple de 41, on utilise l'opérateur modulo: si on divise un nombre par 43 et que le reste est zéro, alors le nombre est multiple de 41. 

Voici le code pour lire un nombre au clavier et tester si il est multiple de 41:

In [2]:
import java.util.Scanner;
Scanner clavier = new Scanner(System.in);

In [3]:
System.out.println("Entrer un nombre:");
int nombre = clavier.nextInt();
if (nombre%41==0){
    System.out.println(nombre +" est multiple de 41");
} else {
    System.out.println(nombre +" n'est pas multiple de 41");
}

Entrer un nombre:
534
534 n'est pas multiple de 41


Pour savoir si un nombre dans la liste au dessus est multiple de 41, on peut tous les tester. Pour cela, on place le test à l'intérieur de la boucle, et on remplace le ```nombre``` dans le test par ```i```, qui prend successivement toutes les valeurs qu'on veut tester:

In [5]:
for (int i=191; i<=207; i++){
    if (i%41==0){
        System.out.println(i +" est multiple de 41");
    } else {
        System.out.println(i +" n'est pas multiple de 41");
    }
}

191 n'est pas multiple de 41
192 n'est pas multiple de 41
193 n'est pas multiple de 41
194 n'est pas multiple de 41
195 n'est pas multiple de 41
196 n'est pas multiple de 41
197 n'est pas multiple de 41
198 n'est pas multiple de 41
199 n'est pas multiple de 41
200 n'est pas multiple de 41
201 n'est pas multiple de 41
202 n'est pas multiple de 41
203 n'est pas multiple de 41
204 n'est pas multiple de 41
205 est multiple de 41
206 n'est pas multiple de 41
207 n'est pas multiple de 41


En lisant ces résultats à l'écran, on peut voir qu'un nombre dans l'intervalle (205) est multiple de 41. Mais dans le contexte d'un programme plus long, on voudrait pouvoir faire d'autres calculs en utilisant ces informations: soit l'information que oui, l'intervalle contient bien un multiple de 41 (information de type booléenne) ou bien l'information que l'élément de la liste qui correspond à la condition est le nombre 205 (information de type numérique).

La première est celle qui a tendance a causer des erreurs.  On définit donc une variable booléenne ```trouvé```, et on voudrait qu'après la boucle ```trouvé``` soit ```true``` si l'intervalle contient un multiple de 41. La première intuition est d'assigner le résultat du "test" dans la boucle à la variable:

In [8]:
boolean trouvé;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé=true;
    } else {
        trouvé=false;
    }
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41!");
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle ne contient pas de multiple de 41!


Que se passe-t-il?

En fait si on détaille bien ce qui se passe, pour chaque valeur de i on effectue le test et on assigne un valeur à ```trouvé```. La valeur de ```trouvé``` après la boucle est donc la valeur obtenue lors du *dernier* test (pour ```i = 207```. Et comme 207 n'est pas multiple de 41, la valeur de ```trouvé``` est ```false```.

La solution ici est la suivante:
Il faut initialiser ```trouvé``` à ```false``` (avant la boucle) et seulement modifier ```trouvé``` *si on trouve ce qu'on cherche*, c'est à dire ici, si un des nombres de l'intervalle est multiple de 41.



In [9]:
boolean trouvé=false;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé=true;
    } 
    // on ne met pas de clause else!
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41!");
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41!


Cette fois, ça fonctionne correctement. Par souci d'efficacité, on peut aussi quitter la boucle une fois que le nombre est trouvé: continuer à parcourir l'intervalle ne changera rien. 

Note: Ceci serait surtout utile si on parcourait une liste très longue; ici on ne verra pas de différence. 

Pour quitter la boucle, on utilise une instruction ```break```:

In [10]:
boolean trouvé=false;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé=true;
        break; // ça y est, on l'a trouvé! on peut quitter la boucle.
    } 
    // on ne met pas de clause else!
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41!");
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41!


Quelques erreurs classiques:
    * la clause ```else```, comme montré au dessus
    * mettre le ```break``` en dehors du ```if```: on quitte alors la boucle dès le premier passage dans la boucle.

On peut maintenant voir comment récupérer la valeur du nombre multiple de 41. Comme pour l'information "trouvé", on va créer une variable pour la recevoir (cette fois un entier), et lui donner une valeur dans la condition ```if``` qui indique qu'une valeur appropriée a été trouvée. Ici on nomme cette variable ```multiple```. La seule difficulté est de donner une valeur initiale à ```multiple```: si on ne trouve pas de multiple, alors la variable ne veut rien dire. Ici on l'initialise à 0. Comme on garde aussi la variable ```trouvé```, on n'aura pas de difficulté à interpréter la variable ```multiple```.

In [12]:
boolean trouvé=false;
int multiple =0;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé = true;
        multiple = i;
        break; // ça y est, on l'a trouvé! on peut quitter la boucle.
    } 
    // on ne met pas de clause else!
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41: "+ multiple);
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41: 205


Parfois, pour simplifier le code (ou dans le contexte d'une fonction, qu'on verra plus tard), on utilise une seule variable pour obtenir la valeur de l'élément cherché, et on utilise une valeur particulière pour le cas où on ne trouve pas d'élément:

In [13]:
int multiple =0;
for (int i=191; i<=207; i++){
    if (i%41==0){
        multiple = i;
        break; // ça y est, on l'a trouvé! on peut quitter la boucle.
    } 
    // on ne met pas de clause else!
}

if (multiple!=0){
    System.out.println("L'intervalle contient bien un multiple de 41: "+ multiple);
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41: 205


Ici on initialise ```multiple``` à zéro, et si cette valeur n'a pas changé on suppose que l'élément cherché n'a pas été trouvé. Mais il faut faire attention: 0 est en fait multiple de 41, et si on cherchait des multiples de 41 dans l'intervalle \[-10, 10\], la valeur 0 serait la valeur du multiple trouvé, et non pas une indication qu'on l'a trouvé. Il faudrait alors utiliser une autre valeur initiale, ou simplement utiliser une deuxième variable comme au-dessus.

### Exercice 5: trouver la racine entière d'un nombre

On veut trouver la racine carrée d'un nombre *n*, si elle est entière. Pour ceci, on va simplement parcourir l'intervalle \[0, n\] et tester chaque nombre pour voir si il donne *n* quand on l'élève au carré. Si on ne trouve pas de racine entière, on affichera un message indiquant cela. Utiliser le squelette de code suivant comme point de départ:

In [ ]:
Scanner clavier = new Scanner(System.in);
int nombre;
do {
    System.out.println("Entrer un entier positif pour obtenir sa racine carrée:");
    nombre = clavier.nextInt();
} while (...); // pour s'assurer que le nombre est positif

int racine = ...;
for ( ... ){ // boucle pour chercher la racine dans l'intervalle [0, nombre]
    
}

if (racine == ...){
    System.out.println("La racine du nombre n'est pas entière");
} else {
    System.out.println("La racine du nombre est "+ racine);
}